File: server/core/parser.go
package core

import (
	"bytes"
	"encoding/binary"
)

func CreateParser(buffer []byte) *Parser {
	var parser = new(Parser)

	parser.MsgBuffer = buffer
	parser.BigEndian = true
	parser.MsgLength = uint32(len(buffer))

	parser.PeerId = parser.ParseDword()
	parser.TaskId = parser.ParseDword()
	parser.MsgType = parser.ParseDword()

	return parser
}

func (p *Parser) ParseByte() []byte {
	var buffer = make([]byte, 1)

	for i := range buffer {
		buffer[i] = 0
	}

	if p.MsgLength >= 1 {

		copy(buffer, p.MsgBuffer[:1])
		p.MsgLength -= 1

		if p.MsgLength == 0 {
			p.MsgBuffer = []byte{}

		} else {
			p.MsgBuffer = p.MsgBuffer[1:]
		}
	}

	return buffer
}

func (p *Parser) ParseBool() bool {
	var integer = make([]byte, 4)

	for i := range integer {
		integer[i] = 0
	}

	if p.MsgLength >= 4 {

		copy(integer, p.MsgBuffer[:4])
		p.MsgLength -= 4

		if p.MsgLength == 0 {
			p.MsgBuffer = []byte{}

		} else {
			p.MsgBuffer = p.MsgBuffer[4:]
		}
	}

	if p.BigEndian {
		return int(binary.BigEndian.Uint32(integer)) != 0

	} else {
		return int(binary.LittleEndian.Uint32(integer)) != 0
	}
}

func (p *Parser) ParseDword() uint32 {
	var buffer = make([]byte, 4)

	for i := range buffer {
		buffer[i] = 0
	}

	if p.MsgLength >= 4 {

		copy(buffer, p.MsgBuffer[:4])
		p.MsgLength -= 4

		if p.MsgLength == 0 {
			p.MsgBuffer = []byte{}

		} else {
			p.MsgBuffer = p.MsgBuffer[4:]
		}
	}

	if p.BigEndian {
		return binary.BigEndian.Uint32(buffer)

	} else {
		return binary.LittleEndian.Uint32(buffer)
	}
}

func (p *Parser) ParseDword64() uint64 {
	var buffer = make([]byte, 8)

	for i := range buffer {
		buffer[i] = 0
	}

	if p.MsgLength >= 8 {

		copy(buffer, p.MsgBuffer[:8])
		p.MsgLength -= 8

		if p.MsgLength == 0 {
			p.MsgBuffer = []byte{}
		} else {
			p.MsgBuffer = p.MsgBuffer[8:]
		}
	}

	if p.BigEndian {
		return binary.BigEndian.Uint64(buffer)

	} else {
		return binary.LittleEndian.Uint64(buffer)
	}
}

func (p *Parser) ParseBytes() []byte {
	var buffer []byte

	if p.MsgLength >= 4 {
		size := p.ParseDword()

		if size != 0 {
			buffer = make([]byte, size)
			copy(buffer, p.MsgBuffer[:size])

			p.MsgLength -= size

			if p.MsgLength == 0 {
				p.MsgBuffer = []byte{}

			} else {
				p.MsgBuffer = p.MsgBuffer[size:]
			}
		}
	}

	return buffer
}

func (p *Parser) ParseWString() string {
	return string(bytes.Trim([]byte(DecodeUTF16(p.ParseBytes())), "\x00"))
}

func (p *Parser) ParseString() string {
	return string(p.ParseBytes())
}

File: rust_server/src/server/parser.rs
use byteorder::{BigEndian, LittleEndian, ByteOrder};
use std::convert::TryInto;

pub struct Parser {
    msg_buffer: Vec<u8>,
    big_endian: bool,
    msg_length: u32,
    peer_id:    u32,
    task_id:    u32,
    msg_type:   u32,
}

impl Parser {
    pub fn create_parser(buffer: Vec<u8>) -> Parser {
        let mut parser = Parser {
            msg_buffer: buffer,
            big_endian: true,
            msg_length: 0,
            peer_id:    0,
            task_id:    0,
            msg_type:   0,
        };

        parser.msg_length   = parser.msg_buffer.len() as u32;
        parser.peer_id      = parser.parse_dword();
        parser.task_id      = parser.parse_dword();
        parser.msg_type     = parser.parse_dword();

        parser
    }

    pub fn parse_byte(&mut self) -> Vec<u8> {
        if self.msg_length >= 1 {
            let buffer = self.msg_buffer[..1].to_vec();

            self.msg_buffer = self.msg_buffer[1..].to_vec();
            self.msg_length -= 1;

            buffer
        } else {
            vec![0; 1]
        }
    }

    pub fn parse_bool(&mut self) -> bool {
        let integer = self.parse_dword();
        integer != 0
    }

    pub fn parse_dword(&mut self) -> u32 {
        if self.msg_length >= 4 {
            let buffer = &self.msg_buffer[..4];

            self.msg_buffer = self.msg_buffer[4..].to_vec();
            self.msg_length -= 4;

            if self.big_endian {
                BigEndian::read_u32(buffer)
            } else {
                LittleEndian::read_u32(buffer)
            }
        } else {
            0
        }
    }

    pub fn parse_dword64(&mut self) -> u64 {
        if self.msg_length >= 8 {
            let buffer = &self.msg_buffer[..8];

            self.msg_buffer = self.msg_buffer[8..].to_vec();
            self.msg_length -= 8;

            if self.big_endian {
                BigEndian::read_u64(buffer)
            } else {
                LittleEndian::read_u64(buffer)
            }
        } else {
            0
        }
    }

    pub fn parse_bytes(&mut self) -> Vec<u8> {
        let size = self.parse_dword() as usize;

        if size > 0 && self.msg_length as usize >= size {
            let buffer = self.msg_buffer[..size].to_vec();

            self.msg_buffer = self.msg_buffer[size..].to_vec();
            self.msg_length -= size as u32;

            buffer
        } else {
            vec![]
        }
    }

    pub fn parse_wstring(&mut self) -> String {
        let bytes = self.parse_bytes();
        let string = decode_utf16(&bytes);

        string.trim_end_matches('\x00').to_string()
    }

    pub fn parse_string(&mut self) -> String {
        String::from_utf8(self.parse_bytes()).unwrap_or_default()
    }
}

fn decode_utf16(bytes: &[u8]) -> String {
    let mut u16_vec = Vec::with_capacity(bytes.len() / 2);

    for chunk in bytes.chunks(2) {
        let arr: [u8; 2] = chunk.try_into().unwrap_or([0, 0]);
        u16_vec.push(u16::from_le_bytes(arr));
    }
    String::from_utf16_lossy(&u16_vec)
}
